sklearn/cluster/k_medoids_.py

```diff
@@ -0,0 +1,292 @@
+            curr_cost = in_cluster_all_costs[
+                np.argmax(cluster_k_idxs == medoid_idxs[k])]
+
+            # Adopt a new medoid if its distance is smaller then the current
+            if min_cost < curr_cost:
+                medoid_idxs[k] = cluster_k_idxs[min_cost_idx]
+
+    def transform(self, X):
+        """Transforms X to cluster-distance space.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = (n_samples, n_features)
+            Data to transform.
+
+        Returns
+        -------
+        X_new : {array-like, sparse matrix}, shape=(n_samples, n_clusters)
+            X transformed in the new space of distances to cluster centers.
+        """
+        X = check_array(X, accept_sparse=['csr', 'csc'])
+        check_is_fitted(self, "cluster_centers_")
+
+        return pairwise_distances(X, Y=self.cluster_centers_,
+                                  metric=self.metric)
+
+    def predict(self, X):
+        """Predict the closest cluster for each sample in X
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = (n_samples, n_features)
+            New data to predict.
+
+        Returns
+        -------
+        labels : array, shape = (n_samples,)
+            Index of the cluster each sample belongs to.
+        """
+        check_is_fitted(self, "cluster_centers_")
+        X = check_array(X, accept_sparse=['csr', 'csc'])
+
+        # Return data points to clusters based on which cluster assignment
+        # yields the smallest distance
+        return pairwise_distances_argmin(X, Y=self.cluster_centers_,
+                                         metric=self.metric)
+
+    def _compute_inertia(self, distances):
+        """Compute inertia of new samples. Inertia is defined as the sum of the
+        sample distances to closest cluster centers.
+
+        Parameters
+        ----------
+        distances : {array-like, sparse matrix}, shape=(n_samples, n_clusters)
+            Distances to cluster centers.
+
+        Returns
+        -------
+        Sum of sample distances to closest cluster centers.
+        """
+
+        # Define inertia as the sum of the sample-distances
+        # to closest cluster centers
+        inertia = np.sum(np.min(distances, axis=1))
+
+        return inertia
+
+    def _initialize_medoids(self, D, n_clusters):
+        """Select initial mediods randomly or heuristically"""
+
+        if self.init == 'random':  # Random initialization
+            # Pick random k medoids as the initial ones.
+            medoids = self.random_state_.choice(len(D), n_clusters)
+        else:  # Initialization by heuristic
+            # Pick K first data points that have the smallest sum distance
+            # to every other point. These are the initial medoids.
+            medoids = np.argpartition(np.sum(D, axis=1),
+                                      range(0, n_clusters))[:n_clusters]
```

Comment:
I don't think you need the range. Just `argpartition(..., n_clusters)` should ensure that the first `n_clusters` elements are the smallest, but doesn't ensure anything about the order within that group.